detector.go

```diff
@@ -556,45 +772,82 @@ type depEntry struct {
 	extraProvides []Buildpack
 }
 
-type depMap map[string]depEntry
+type depMap map[depKey]depEntry
 
-func newDepMap(trial detectTrial) depMap {
+var anyMixinKey depKey = depKey{name: "*", mixin: true, any: true}
+
+func newDepMap(trial detectTrial, stage stageName) (depMap, error) {
 	m := depMap{}
 	for _, option := range trial {
 		for _, p := range option.Provides {
-			m.provide(option.Buildpack, p)
+			if p.Mixin && !option.Buildpack.Privileged {
+				return nil, fmt.Errorf(errInvalidProvidesBuildpack, option.Buildpack.String())
+			}
+
+			for _, prv := range p.toStageProvides() {
+				m.provide(stage, option.Buildpack, prv)
+			}
 		}
 		for _, r := range option.Requires {
-			m.require(option.Buildpack, r)
+			for _, req := range r.toStageRequires() {
+				m.require(stage, option.Buildpack, req)
+			}
 		}
 	}
-	return m
+	return m, nil
 }
 
-func (m depMap) provide(bp Buildpack, provide Provide) {
-	entry := m[provide.Name]
+func (m depMap) provide(stage stageName, bp Buildpack, provide Provide) {
+	if !bp.Privileged && provide.stage() != stage {
+		return
+	}
+
+	// if it is not a match, use the full name to not match when required
```

Comment:
Is it possible you could elaborate on this comment?

Reply:
This will probably be hard to explain and may take me a few attempts. Let's set the stage a bit. `detect` passing AND providing a build plan means allows buildpacks to skip in addition to pass/fail via exit code alone. If a buildpack is optional and the buildpack writes a `provides` of `libpq` - we skip the buildpack but don't fail detection.

This concept becomes tricker in stackpacks. Because a single detection may produce a buildplan like this:
```
[[provides]]
name = run:libpq
mixin = true
```
Let's say a user buildpack has
```
[[requires]]
name = run:libpq
mixin = true
```
The mere _existence_ of a build plan that `provides` something means we should NOT solely use the detect exit code. In this case I want this stackpack to be `skipped` during `build` phase and to not be skipped during `run` phase. 

Visualized, the logs might look something like...
```
...
Resolving plan... (try #1)
skip: Stackpack/A@v1 [build] - not required
pass: Stackpack/A@v1 [run]
pass: userbuildpack/z@v1
...
```

If I were to simply remove the non-stage matching `provide`/`requires` entirely when doing the dependency resolution per-stage, the stackpack would end up executing during `build` phase. This is because if the dependency map doesn't contain anything, it is assumed that there was no build plan essentially and it relies on the exit code alone. The reverse of this problem is more costly. I 100% don't want to try and extend the run image if there is nothing to be done on the image.

So anyway...there is probably a better way to fix this. The way I chose to do it for now is to allow stage specific mixins to exist in the map for stages not being processed. It's a silly way to fix not having an empty depMap.

`should not require extending run image for build stage mixins` is one of the few tests that maybe help explain this a bit more.